src/gradient_descent.c

```diff
@@ -0,0 +1,380 @@
+      inflexion = (bracket.c) + GOLD * (bracket.c - bracket.b);
+      fInflexion = (*func)(f, inflexion, initialPoint, direction, n);
+    }
+
+    // Check if u is between c and the limit
+    else if ((bracket.c - inflexion) * (inflexion - ulim) > 0.0) {
+
+      // Compute function value at u
+      fInflexion = (*func)(f, inflexion, initialPoint, direction, n);
+
+      // Check if we need to continue looking after u
+      if (fInflexion < fc) {
+        // Magnify u and proceed to the next iteration
+        bracket.b = bracket.c;
+        bracket.c = inflexion;
+        inflexion = bracket.c + GOLD * (bracket.c - bracket.b);
+
+        fb = fc;
+        fc = fInflexion;
+        fInflexion = (*func)(f, inflexion, initialPoint, direction, n);
+      }
+    }
+
+    // If inflexion point is further than the limit
+    else if ((inflexion - ulim) * (ulim - bracket.c) >= 0.0) {
+      inflexion = ulim;
+      fInflexion = (*func)(f, inflexion, initialPoint, direction, n);
+    }
+
+    // Reject parabolic interpolation
+    // Magnify and proceed to next iteration
+    else {
+      inflexion = (bracket.c) + GOLD * (bracket.c - bracket.b);
+      fInflexion = (*func)(f, inflexion, initialPoint, direction, n);
+    }
+
+    // Eliminate oldest point and continue
+    bracket.a = bracket.b;
+    bracket.b = bracket.c;
+    bracket.c = inflexion;
+
+    fa = fb;
+    fb = fc;
+    fc = fInflexion;
+  }
+
+  return bracket;
+}
+
+/// @brief Applies linear search to find the minimum of a function in a given
+/// direction
+/// @param point Initial point to base the movement of. Will contain the minimum
+/// after the function executes
+/// @param direction Direction in which to move
+/// @param n Number of dimensions of the original function
+/// @param func Function we want to minimize
+/// @param tol Tolerance for the minimum estimate. Should be no smaller then the
+/// square-root of the machine floating point precision
+/// @return Value of the original function at the minimum found
+static gradient_real lineSearch(gradient_real point[],
+                                gradient_real direction[], int n, function func,
+                                gradient_real tol) {
+
+  // Initially bracket the minimum between 3 points
+  Bracket bracket = bracketMinimum(oneDimension, func, point, direction, n);
+
+  // Find the minimum on the line given by the direction vector
+  gradient_real xmin = 0.0;
+  gradient_real value =
+      brent(&xmin, bracket, oneDimension, func, point, direction, n, tol);
+
+  // Move point to the minimum
+  for (int i = 0; i < n; i++) {
+    point[i] += direction[i] * xmin;
+  }
+
+  return value;
+}
+
+gradient_real gradient_descent(function func, derivative dfunc,
+                               gradient_real guess[], int n, gradient_real tol,
+                               int itMax) {
+  // Gradient of the function
+  gradient_real *gradient = malloc(n * sizeof(gradient_real));
```

Comment:
Malloc but no free (That looks like a memory leak to me ;) )? I think we can allocate those variables on the stack.